File: Etapa3-Construcao/Stack2/hotel-booking-java/src/main/java/com/hotel/booking/valueobjects/Periodo.java
package com.hotel.booking.valueobjects;

import jakarta.persistence.Embeddable;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import java.time.LocalDate;
import java.time.format.DateTimeFormatter;

@Embeddable
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Periodo {
    private LocalDate checkIn;
    private LocalDate checkOut;

    public Periodo(LocalDate checkIn, LocalDate checkOut) {
        this.checkIn = checkIn;
        this.checkOut = checkOut;
        validarPeriodo();
    }

    private void validarPeriodo() {
        if (checkOut == null || checkIn == null) {
            throw new IllegalArgumentException("Datas de check-in e check-out são obrigatórias");
        }

        if (checkOut.isBefore(checkIn) || checkOut.isEqual(checkIn)) {
            throw new IllegalArgumentException("Check-out deve ser posterior ao check-in");
        }

        if (checkIn.isBefore(LocalDate.now())) {
            throw new IllegalArgumentException("Check-in não pode ser no passado");
        }

        long dias = getNumeroNoites();
        if (dias > 30) {
            throw new IllegalArgumentException("Período de estadia não pode exceder 30 dias");
        }
    }

    public long getNumeroNoites() {
        return java.time.temporal.ChronoUnit.DAYS.between(checkIn, checkOut);
    }

    public boolean contem(LocalDate data) {
        return !data.isBefore(checkIn) && data.isBefore(checkOut);
    }

    public boolean sobrepoe(Periodo outro) {
        return this.checkIn.isBefore(outro.checkOut) && this.checkOut.isAfter(outro.checkIn);
    }

    public boolean ehValidoParaReserva() {
        return checkIn.isAfter(LocalDate.now().minusDays(1)) && getNumeroNoites() > 0;
    }

    public String getPeriodoFormatado() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy");
        return checkIn.format(formatter) + " a " + checkOut.format(formatter);
    }

    public boolean ehFuturo() {
        return checkIn.isAfter(LocalDate.now());
    }

    public long diasAteCheckIn() {
        LocalDate hoje = LocalDate.now();
        return java.time.temporal.ChronoUnit.DAYS.between(hoje, checkIn);
    }
}

File: Etapa3-Construcao/Stack1/hotel-booking-system/src/domain/value-objects/Periodo.ts
// Value Object: Período de Estadia
export class Periodo {
  constructor(
    public readonly checkIn: Date,
    public readonly checkOut: Date
  ) {
    this.validarPeriodo();
  }

  // Validação do período
  private validarPeriodo(): void {
    if (this.checkOut <= this.checkIn) {
      throw new Error('Check-out deve ser posterior ao check-in');
    }

    if (this.checkIn < new Date()) {
      throw new Error('Check-in não pode ser no passado');
    }

    // Verifica se o período é muito longo (ex: mais de 30 dias)
    const dias = this.getNumeroNoites();
    if (dias > 30) {
      throw new Error('Período de estadia não pode exceder 30 dias');
    }
  }

  // Calcula o número de noites
  getNumeroNoites(): number {
    const msPorDia = 24 * 60 * 60 * 1000;
    return Math.ceil((this.checkOut.getTime() - this.checkIn.getTime()) / msPorDia);
  }

  // Verifica se uma data está dentro do período
  contem(data: Date): boolean {
    return data >= this.checkIn && data < this.checkOut;
  }

  // Verifica se há sobreposição com outro período
  sobrepoe(outro: Periodo): boolean {
    return this.checkIn < outro.checkOut && this.checkOut > outro.checkIn;
  }

  // Verifica se o período é válido para reserva
  ehValidoParaReserva(): boolean {
    const hoje = new Date();
    hoje.setHours(0, 0, 0, 0);
    
    return this.checkIn >= hoje && this.getNumeroNoites() > 0;
  }

  // Formata o período para exibição
  getPeriodoFormatado(): string {
    const formatarData = (data: Date): string => {
      return data.toLocaleDateString('pt-BR', {
        day: '2-digit',
        month: '2-digit',
        year: 'numeric'
      });
    };

    return `${formatarData(this.checkIn)} a ${formatarData(this.checkOut)}`;
  }

  // Cria uma cópia com novas datas
  atualizar(checkIn?: Date, checkOut?: Date): Periodo {
    return new Periodo(
      checkIn ?? this.checkIn,
      checkOut ?? this.checkOut
    );
  }

  // Converte para objeto JSON
  toJSON() {
    return {
      checkIn: this.checkIn.toISOString(),
      checkOut: this.checkOut.toISOString(),
      numeroNoites: this.getNumeroNoites()
    };
  }

  // Cria instância a partir de objeto JSON
  static fromJSON(dados: any): Periodo {
    return new Periodo(
      new Date(dados.checkIn),
      new Date(dados.checkOut)
    );
  }

  // Cria período a partir de strings de data
  static fromStrings(checkInStr: string, checkOutStr: string): Periodo {
    const checkIn = new Date(checkInStr);
    const checkOut = new Date(checkOutStr);
    
    return new Periodo(checkIn, checkOut);
  }

  // Verifica se o período é futuro
  ehFuturo(): boolean {
    return this.checkIn > new Date();
  }

  // Calcula dias até o check-in
  diasAteCheckIn(): number {
    const hoje = new Date();
    hoje.setHours(0, 0, 0, 0);
    const checkIn = new Date(this.checkIn);
    checkIn.setHours(0, 0, 0, 0);
    
    const msPorDia = 24 * 60 * 60 * 1000;
    return Math.ceil((checkIn.getTime() - hoje.getTime()) / msPorDia);
  }
}
